django/db/models/expressions.py

```diff
@@ -292,16 +292,22 @@ def resolve_expression(
         """
         c = self.copy()
         c.is_summary = summarize
-        c.set_source_expressions(
-            [
-                (
-                    expr.resolve_expression(query, allow_joins, reuse, summarize)
-                    if expr
-                    else None
-                )
-                for expr in c.get_source_expressions()
-            ]
-        )
+        source_expressions = [
+            (
+                expr.resolve_expression(query, allow_joins, reuse, summarize)
+                if expr is not None
+                else None
+            )
+            for expr in c.get_source_expressions()
+        ]
+        if not self.allows_composite_expressions and any(
+            isinstance(expr, ColPairs) for expr in source_expressions
+        ):
+            raise ValueError(
+                f"{self.__class__.__name__} expression does not support "
```

Comment:
I added the word _expression_ here as the error message read strangely for `Case` and `When` otherwise.